src/Core/Managed/Shared/Extensibility/MetricManager.cs

```diff
@@ -0,0 +1,285 @@
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MetricManager"/> class.
+        /// </summary>
+        /// <param name="client">Telemetry client to use to output aggregated metric data.</param>
+        public MetricManager(TelemetryClient client)
+        {
+            this.telemetryClient = client ?? new TelemetryClient();
+            this.metricDictionary = new ConcurrentDictionary<Metric, SimpleMetricStatisticsAggregator>();
+
+            this.lastSnapshotStartDateTime = DateTimeOffset.UtcNow;
+
+            this.snapshotTimer = new TaskTimer() { Delay = GetWaitTime() };
+            this.snapshotTimer.Start(this.SnapshotAndReschedule);
+        }
+
+        /// <summary>
+        /// Gets a list of metric processors associated
+        /// with this instance of <see cref="MetricManager"/>.
+        /// </summary>
+        internal IList<IMetricProcessor> MetricProcessors
+        {
+            get
+            {
+                TelemetryConfiguration config = this.telemetryClient.TelemetryConfiguration;
+
+                return config.MetricProcessors;
+            }
+        }
+
+        /// <summary>
+        /// Creates metric.
+        /// </summary>
+        /// <param name="name">Name of the metric.</param>
+        /// <param name="dimensions">Optional dimensions.</param>
+        /// <returns>Metric instance.</returns>
+        public Metric CreateMetric(string name, IDictionary<string, string> dimensions = null)
+        {
+            if (this.disposed)
+            {
+                throw new ObjectDisposedException(nameof(MetricManager));
+            }
+
+            return new Metric(this, name, dimensions);
+        }
+
+        /// <summary>
+        /// Flushes the in-memory aggregation buffers.
+        /// </summary>
+        public void Flush()
+        {
+            if (this.disposed)
+            {
+                throw new ObjectDisposedException(nameof(MetricManager));
+            }
+
+            this.Flush(false);
+        }
+
+        /// <summary>
+        /// Disposes the object.
+        /// </summary>
+        public void Dispose()
+        {
+            this.disposed = true;
+
+            if (this.snapshotTimer != null)
+            {
+                this.snapshotTimer.Dispose();
+                this.snapshotTimer = null;
+            }
+
+            this.Flush(true);
+        }
+
+        internal SimpleMetricStatisticsAggregator GetStatisticsAggregator(Metric metric)
+        {
+            if (this.disposed)
+            {
+                throw new ObjectDisposedException(nameof(MetricManager));
+            }
+
+            return this.metricDictionary.GetOrAdd(metric, (m) => { return new SimpleMetricStatisticsAggregator(); });
+        }
+
+        /// <summary>
+        /// Calculates wait time until next snapshot of the aggregators.
+        /// </summary>
+        /// <returns>Wait time.</returns>
+        private static TimeSpan GetWaitTime()
```

Comment:
This logic is really hard to read, and I think it is also not correct. if your aggregation period is >= 1 minute, your 3 seconds rule will never apply. And is it is small, then you may still end up in the past.
Consider this as an alternative:

`private static DateTimeOffset GetNextSnapshitTime(DateTimeOffset prevTime, TimeSpan aggregationPeriod)
        { 
            DateTimeOffset exactWaitUntilTime = prevTime + aggregationPeriod;

            // Below we have logic that rounds the next snapshot event to a sertain number of seconds past a minute.
            // This would break down if the aggregation period was smaller than a minute. 
            // So, for small aggegation periods we are ot rounding.
            if (aggregationPeriod < TimeSpan.FromMinutes(1))
            {
                return exactWaitUntilTime;
            }

            // We want to snapshot at 1 sec past full minute.
            // Why??
            const int pastMinSecs = 1;
            DateTimeOffset roundedWaitUntilTime = new DateTimeOffset(exactWaitUntilTime.Year, exactWaitUntilTime.Month, exactWaitUntilTime.Day,
                                                                        exactWaitUntilTime.Hour, exactWaitUntilTime.Minute, pastMinSecs,
                                                                        exactWaitUntilTime.Offset);
            // We round to the closest minute boundry (Attention, adjust this condition if NOT (0 <= pastMinSecs < 30) !! )
            if (exactWaitUntilTime.Second > (30 + pastMinSecs))
            {
                roundedWaitUntilTime = roundedWaitUntilTime.AddMinutes(1);
            }

            return roundedWaitUntilTime;
        }`